File: src/test/java/io/mfedirko/RepositoryTestHelpers.java
package io.mfedirko;

import lombok.experimental.UtilityClass;
import org.assertj.core.api.Condition;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

@UtilityClass
public class RepositoryTestHelpers {
    public static final ZoneId TZ_CHICAGO = ZoneId.of("America/Chicago");

    public static <T> Condition<? super List<? extends T>> sortedDescending(Function<T, LocalDateTime> timestampGetter) {
        return new Condition<>(items -> {
            List<? extends T> sorted = new ArrayList<>(items);
            sorted.sort(Comparator.comparing(timestampGetter).reversed());
            return sorted.equals(items);
        }, "List should be sorted descending by timestamp");
    }


    public static <T> Condition<T> withinDate(LocalDate date, Function<T, LocalDateTime> timestampGetter) {
        LocalDateTime from = date.atStartOfDay(TZ_CHICAGO).toLocalDateTime();
        LocalDateTime to = date.atTime(23,59,59).atZone(TZ_CHICAGO).toLocalDateTime();
        return withinRange(from, to, timestampGetter);
    }

    public static <T> Condition<T> withinYear(int year, Function<T, LocalDateTime> timestampGetter) {
        LocalDateTime from = LocalDate.of(year, 1, 1).atStartOfDay(TZ_CHICAGO).toLocalDateTime();
        LocalDateTime to = LocalDate.of(year, 12, 31).atTime(23,59,59).atZone(TZ_CHICAGO).toLocalDateTime();
        return withinRange(from, to, timestampGetter);
    }

    public static <T> Condition<T> withinRange(LocalDateTime from, LocalDateTime to, Function<T, LocalDateTime> timestampGetter) {
        return new Condition<>(item -> timestampGetter.apply(item).isAfter(from)
                && timestampGetter.apply(item).isBefore(to),
                "Should be within range %s - %s.".formatted(from, to));
    }
}

File: src/test/kotlin/io/mfedirko/RepositoryTestHelpers.kt
package io.mfedirko

import org.assertj.core.api.Condition
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.function.Function

object RepositoryTestHelpers {
    val TZ_CHICAGO = ZoneId.of("America/Chicago")

    fun <T> sortedDescending(timestampGetter: (T) -> LocalDateTime): Condition<in List<T>> {
        return Condition({ items: List<T> ->
            val sorted: List<T> = ArrayList(items).sortedWith(compareByDescending(timestampGetter))
            sorted == items
        }, "List should be sorted descending by timestamp")
    }

    fun <T> withinDate(date: LocalDate, timestampGetter: (T) -> LocalDateTime): Condition<T> {
        val from = date.atStartOfDay(TZ_CHICAGO).toLocalDateTime()
        val to = date.atTime(23, 59, 59).atZone(TZ_CHICAGO).toLocalDateTime()
        return withinRange(from, to, timestampGetter)
    }

    fun <T> withinYear(year: Int, timestampGetter: (T) -> LocalDateTime): Condition<T> {
        val from = LocalDate.of(year, 1, 1).atStartOfDay(TZ_CHICAGO).toLocalDateTime()
        val to = LocalDate.of(year, 12, 31).atTime(23, 59, 59).atZone(TZ_CHICAGO).toLocalDateTime()
        return withinRange(from, to, timestampGetter)
    }

    fun <T> withinRange(
        from: LocalDateTime?,
        to: LocalDateTime?,
        timestampGetter: Function<T, LocalDateTime>
    ): Condition<T> {
        return Condition(
            { item: T ->
                (timestampGetter.apply(item).isAfter(from)
                        && timestampGetter.apply(item).isBefore(to))
            },
            "Should be within range $from - $to."
        )
    }
}